File: src/observers/BudgetObserver.js
/**
 * Observer: следит за новыми расходами и пишет предупреждение,
 * если категория вышла за месячный бюджет.
 */
class BudgetObserver {
  constructor({ eventBus, transactionRepo, budgetRepo, alertRepo }) {
    this.transactionRepo = transactionRepo;
    this.budgetRepo = budgetRepo;
    this.alertRepo = alertRepo;
    eventBus.subscribe("transaction:created", (tx) => this.onTransaction(tx));
  }

  onTransaction(tx) {
    if (tx.type !== "expense") return;

    const budget = this.budgetRepo.findByCategoryId(tx.categoryId);
    if (!budget) return;

    const date = new Date(tx.date);
    const monthKey = `${date.getFullYear()}-${date.getMonth() + 1}`;
    const spent = this.transactionRepo
      .findByMonth(date.getFullYear(), date.getMonth())
      .filter((item) => item.type === "expense" && item.categoryId === tx.categoryId)
      .reduce((sum, item) => sum + item.amount, 0);

    if (spent <= budget.limit) return;

    this.alertRepo.addUnique({
      id: crypto.randomUUID(),
      categoryId: tx.categoryId,
      monthKey,
      spent,
      limit: budget.limit,
      message: `Бюджет превышен: потрачено ${spent} ₽ из ${budget.limit} ₽`,
      createdAt: new Date().toISOString(),
    });
  }
}

module.exports = BudgetObserver;

File: app/src/main/java/com/finmate/app/domain/observer/BudgetObserver.kt
package com.finmate.app.domain.observer

import com.finmate.app.data.model.Alert
import com.finmate.app.data.model.Transaction
import com.finmate.app.data.model.TransactionType
import com.finmate.app.data.repository.FinanceRepository
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.util.UUID

/** Observer: предупреждение при превышении бюджета. */
class BudgetObserver(
    eventBus: EventBus,
    private val repository: FinanceRepository
) {
    init {
        eventBus.subscribe("transaction:created") { payload ->
            if (payload is Transaction) onTransaction(payload)
        }
    }

    private fun onTransaction(tx: Transaction) {
        if (tx.type != TransactionType.EXPENSE) return
        val budget = repository.getBudgets().find { it.categoryId == tx.categoryId } ?: return
        val date = OffsetDateTime.parse(tx.date).withOffsetSameInstant(ZoneOffset.UTC)
        val monthKey = "${date.year}-${date.monthValue}"
        val spent = repository.transactionsForMonth(date.year, date.monthValue)
            .filter { it.type == TransactionType.EXPENSE && it.categoryId == tx.categoryId }
            .sumOf { it.amount }
        if (spent <= budget.limit) return

        repository.addAlertUnique(
            Alert(
                id = UUID.randomUUID().toString(),
                categoryId = tx.categoryId,
                monthKey = monthKey,
                spent = spent,
                limit = budget.limit,
                message = "Бюджет превышен: потрачено ${spent.toInt()} ₽ из ${budget.limit.toInt()} ₽",
                createdAt = java.time.Instant.now().toString()
            )
        )
    }
}
